lib/bolt_server/transport_app.rb

```diff
@@ -73,6 +73,38 @@ def run_command(target, body)
       @executor.run_command(target, command)
     end
 
+    def upload_file(target, body)
+      error = validate_schema(@schemas["action-upload_file"], body)
+      return [400, error.to_json] unless error.nil?
+
+      files = body['files']
+      destination = body['destination']
+      job_id = body['job_id']
+      cache_dir = @file_cache.create_cache_dir(job_id.to_s)
+      FileUtils.mkdir_p(cache_dir)
+      files.each do |file|
+        relative_path = file['relative_path']
+        uri = file['uri']
+        sha256 = file['sha256']
+        kind = file['kind']
+        path = File.join(cache_dir, relative_path)
+        if kind == 'file'
+          # The parent should already be created by `directory` entries,
+          # but this is to be on the safe side.
+          parent = File.dirname(path)
+          FileUtils.mkdir_p(parent)
+          @file_cache.download_file(path, sha256, uri)
```

Comment:
This all looks great. It looks like for file downloads we might be able to take advantage of file checks from the file_cache object/class to see if the file has already been downloaded: https://github.com/puppetlabs/bolt/blob/master/lib/bolt_server/file_cache.rb#L106, which could save a bunch of requests to puppetserver as this endpoint will be called multiple times for the same job.

Reply:
Doesn't `download_file` already run `check_file` as a first step? So if it's already in the cache, it won't hit the endpoint?

Reply:
```
    def download_file(file_path, sha, uri)
      if check_file(file_path, sha)
        @logger.debug("File was downloaded while queued: #{file_path}")
        return file_path
      end
      ...
```

Reply:
Oh, hehe, yes it does do that, nevermind :)